Assets/BossRoom/Scripts/Client/Game/Action/AoeActionInput.cs

```diff
@@ -48,7 +49,13 @@ void Update()
             m_InRangeVisualization.SetActive(isInRange);
             m_OutOfRangeVisualization.SetActive(!isInRange);
 
-            if (Input.GetMouseButtonUp(0))
+            // wait for the player to click down and then release the mouse button before actually taking the input
+            if (Input.GetMouseButtonDown(0))
+            {
+                m_ReceivedMouseDownEvent = true;
+            }
+
+            if (Input.GetMouseButtonUp(0) && m_ReceivedMouseDownEvent)
```

Comment:
Why not just MouseButtonUp? Why do you need to make sure down was pressed first?

Reply:
*context (flow of action input)*

* UIHUDButton for "Rain of Arrows" is pressed, and OnPointerDownEvent is fired.
* HeroActionBar.OnButtonClickedDown is called, and ClientInputSender.RequestAction is invoked
* in next Fixed Update, ClientInputSender creates an AOEActionInput to manage the input.
* At this point, the user's finger is still down (from his press on the button). In the old code, he would have to drag his mouse off the button to the spot where he wanted the arrow rain, and then release it, triggering the MouseUp behavior in the AOEActionInput.
* In the new code, he can release his finger, since the AOEActionInput itself requires a full button click to fire the skill.

*question* 
This logic really starts in the UIHUDButton and HeroActionBar, but the same question you raised could be applied to this system as well: why can't HeroActionBar only fire the RequestAction on mouse up? Then the AOEActionInput's logic wouldn't have to change.

*answer*
The answer is that the HeroActionBar system needs to work for a bunch of different skills, and some of them are "charged" skills, where you hold down a button to charge up an effect. These actions must be Requested on mouse down, not mouse up. This was the change that kept things contained to the AOEInputAction, without having to rethink either the action flow, or other skills.